File: week5/hanhee/boj_tsp_2098.cpp
#include <iostream>
using namespace std;
 
const int INF = 987654321;
int N;
int W[17][17];
int dp[16][1<<16];
 
int TSP(int curr, int visited){
    int result;
    int ret = dp[curr][visited];
    if(ret!=0)
        return ret;
    if(visited==((1<<N)-1)){
        if(W[curr][0]!=0)
            return W[curr][0];
        return INF;
    }
 
    ret=INF;
    for(int i=0; i<N;i++){
        if(W[curr][i]==0 || (visited&(1<<i))) 
        ret = min(ret,W[curr][i]+TSP(i,visited|1<<i)); 
    }
    dp[curr][visited]=ret;
    return ret;
}
 
int main(void){
    cin>>N;
    for(int i=0; i<N; i++)
        for(int j=0; j<N; j++)
            cin>>W[i][j];
 
    cout<<TSP(0,1);
}

File: week6/youngjin/boj_tsp_2098.py
import sys
n = int(sys.stdin.readline())
w = [list(map(int, sys.stdin.readline().split())) for _ in range(n)]
INF = sys.maxsize
cache = [[INF]*(1<<n) for _ in range(n)] # 동적 계획법을 위한 캐시 초기화

def tsp(cur_node, visited):
    # 모든 도시를 방문한 경우
    if visited == (1 << n) - 1: 
        return w[cur_node][0] or INF # 마지막도시에서 출발 도시로 돌아가는 비용 반환, 돌아갈 수 없으면 INF 반환(해당 경로는 최소 비용이 될 수 없음)

    # 남은 구간이 이전 탐색과 중복(중복 호출), 계산 생략하고 바로 값을 반환
    if cache[cur_node][visited] != INF: 
        return cache[cur_node][visited]

    min_cost = INF
    for next_node in range(1, n):
        # 방문하지 않았으면서 방문 가능한 도시인 경우
        if not visited & (1 << next_node) and w[cur_node][next_node] != 0: 
            min_cost = min(min_cost, tsp(next_node, visited | (1 << next_node))+w[cur_node][next_node]) # 도시방문, 현재 최소 비용과 비교하여 최소비용을 유지 또는 갱신

    cache[cur_node][visited] = min_cost # 현재 도시, 현재까지 방문한 모든 도시인 경우에서의 최소 비용 저장
    return min_cost

print(tsp(0, 1<<0)) # 시작점을 어디로 잡아도 순환하므로 동일 경로 지남. 시작점 0으로 지정
